src/informo-feeder/poller/poller.go

```diff
@@ -28,143 +29,197 @@ import (
+
+					time.Sleep(time.Duration(feed.PollInterval) * time.Second)
+
+					continue
+				} else if err != nil {
 					logrus.Panic(err)
 				}
 
-				if err = p.sendMatrixEventFromItem(feed, content, item); err != nil {
-					logrus.Panic(err)
-				}
+				// Wait between sending two events in order not to get rate
+				// limited.
+				time.Sleep(500 * time.Millisecond)
 			}
-			p.updateLatestPosition(feed.URL, item)
 
-			time.Sleep(500 * time.Millisecond)
 		}
+
+		// Perform the clear + refill here, this way we don't risk the feeder being
+		// interupted while sending events and restarting with an empty DB for the
+		// feed.
+		if err = p.db.ClearItemsForFeed(feed.Identifier); err != nil {
+			logrus.Panic(err)
+		}
+
+		// Save all items retrieved in this iteration.
+		for _, item := range f.Items {
+			if err = p.db.SaveItem(feed.Identifier, item.Link); err != nil {
+				panic(err)
+			}
+		}
+
+		// Wait before jumping to the next iteration.
+		time.Sleep(time.Duration(feed.PollInterval) * time.Second)
 	}
 }
 
-func (p *Poller) getDurationBeforePoll(feed config.Feed) (timeToPoll time.Duration, err error) {
-	latestPoll, _, err := p.getLatestPosition(feed.URL)
-	if err != nil {
-		return
+// prepareThenSend checks if any HTML could be found in the item (if there is a
+// content, it's always HTML, if not, checks if HTML could be found in the item's
+// description), in which case it will replace media links (with mxc:// URLs)
+// in the item's HTML, then send it to Matrix.
+// Returns an error if no HTML could be found, or if replacing medias failed.
+func (p *Poller) prepareThenSend(feed config.Feed, item *gofeed.Item) error {
+	// Look for HTML content.
+	var content string
+	if len(item.Content) > 0 {
+		// If there's a content, it's always HTML.
+		content = item.Content
+	} else if len(item.Description) > 0 {
+		// If there's a description, check if it contains HTML.
+		if htmlRegexp.MatchString(item.Description) {
+			content = item.Description
+		} else {
+			return errNoHTML
+		}
 	}
 
-	delta := time.Now().Unix() - latestPoll.Unix()
+	logrus.WithFields(logrus.Fields{
+		"title":         item.Title,
+		"publishedDate": item.PublishedParsed.String(),
+	}).Info("Got a new item")
 
-	timeToPoll = time.Duration(feed.PollInterval - delta)
-
-	if timeToPoll < 0 {
-		timeToPoll = 0
+	// Replace media links with mxc:// URLs.
+	if err := p.replaceMedias(&content); err != nil {
+		return err
 	}
 
-	return
+	// Create and send a Matrix event for this item.
+	return p.sendMatrixEventFromItem(feed, content, item)
 }
 
-func (p *Poller) getLatestPosition(
-	feedURL string,
-) (latestPoll time.Time, latestItem time.Time, err error) {
-	lp, li, err := p.db.GetPollerStatusForFeed(feedURL)
-	if err != nil {
-		return
+// itemKnown checks whether an item with a given URL was found in the last poll's
+// results for a given feed.
+func (p *Poller) itemKnown(feedIdentifier string, itemURL string) (known bool) {
```

Comment:
should be optimized
`p.lastPollResults[feedIdentifier]` could be made a hashmap / dictionnary / other so you don't have to loop over each element every time
or at least break if `known == true`